File: packages/senri_ffi_js/src/struct.ts
/*
 * Copyright (c) 2026 TT23XR Studio
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { computeLayout, FieldInfo } from './types/utils';
import { Pointer } from './pointer';

export interface StructClass {
  new (init?: Record<string, any>): StructInstance;
  readonly sizeof: number;
  readonly align: number;
  fromPointer(ptr: any): StructInstance;
  fields: Record<string, any>;
  packed?: number;
}

export interface StructInstance {
  _buffer: ArrayBuffer;
  _view: DataView;
  _fields: FieldInfo[];
  readonly ptr: Pointer;
  toPointer(): Pointer;
  [key: string]: any;
}

let _allocNative: ((size: number) => any) | null = null;
let _freeNative: ((ptr: any) => void) | null = null;

export function setStructAlloc(alloc: (size: number) => any, free: (ptr: any) => void): void {
  _allocNative = alloc;
  _freeNative = free;
}

export function createStruct(fields: Record<string, any>, options?: { packed?: number }): any {
  const packed = options && options.packed;
  const { fieldInfos, totalSize, maxAlign } = computeLayout(fields, packed);

  class StructInstance {
    _buffer: ArrayBuffer;
    _view: DataView;
    _fields: FieldInfo[];

    static get sizeof(): number { return totalSize; }
    static get align(): number { return maxAlign; }

    static fromPointer(ptr: any): StructInstance {
      const inst = Object.create(StructInstance.prototype) as StructInstance;
      inst._buffer = new ArrayBuffer(totalSize);
      inst._view = new DataView(inst._buffer);
      inst._fields = fieldInfos;
      const srcPtr = ptr instanceof Pointer ? ptr : new Pointer(ptr);
      if (srcPtr._data && srcPtr._data.__buf) {
        const src = new Uint8Array(srcPtr._data.__buf);
        const dst = new Uint8Array(inst._buffer);
        dst.set(src.slice(0, totalSize));
      }
      return inst;
    }

    constructor(init?: Record<string, any>) {
      this._buffer = new ArrayBuffer(totalSize);
      this._view = new DataView(this._buffer);
      this._fields = fieldInfos;
      new Uint8Array(this._buffer).fill(0);
      if (init) {
        for (const key of Object.keys(init)) {
          if (key in this) (this as any)[key] = (init as any)[key];
        }
      }
    }

    get ptr(): Pointer {
      return new Pointer({ __buf: this._buffer, __ptr: 0n, __size: totalSize });
    }

    toPointer(): Pointer {
      const native = _allocNative ? _allocNative(totalSize) : null;
      if (native && native.__buf) {
        const dst = new Uint8Array(native.__buf);
        dst.set(new Uint8Array(this._buffer));
        return new Pointer(native);
      }
      return this.ptr;
    }
  }

  for (const fi of fieldInfos) {
    const { name, offset, type, size } = fi;
    Object.defineProperty(StructInstance.prototype, name, {
      get(this: StructInstance) { return readValue(this._view, offset, type, size); },
      set(this: StructInstance, v: any) { writeValue(this._view, offset, type, size, v); },
      enumerable: true,
      configurable: false,
    });
  }

  (StructInstance.prototype as any).__senri_type = 'struct';
  (StructInstance as any).fields = fields;
  (StructInstance as any).packed = packed;

  return StructInstance as unknown as StructClass;
}

function readValue(view: DataView, offset: number, type: any, _size: number): any {
  if (typeof type === 'string') {
    switch (type) {
      case 'int8':    return view.getInt8(offset);
      case 'uint8':   return view.getUint8(offset);
      case 'int16':   return view.getInt16(offset, true);
      case 'uint16':  return view.getUint16(offset, true);
      case 'int32':   return view.getInt32(offset, true);
      case 'uint32':  return view.getUint32(offset, true);
      case 'int64':   return view.getBigInt64(offset, true);
      case 'uint64':  return view.getBigUint64(offset, true);
      case 'float32': return view.getFloat32(offset, true);
      case 'float64': return view.getFloat64(offset, true);
      case 'pointer': case 'cstring': return view.getBigUint64(offset, true);
      default: return 0;
    }
  }
  if (type && type.__senri_type === 'struct' && type.constructor) {
    const cls = type.constructor || type;
    const buf = view.buffer.slice(view.byteOffset + offset, view.byteOffset + offset + _size);
    return cls.fromPointer(new Pointer(buf));
  }
  return 0;
}

function writeValue(view: DataView, offset: number, type: any, _size: number, value: any): void {
  if (typeof type === 'string') {
    switch (type) {
      case 'int8':    view.setInt8(offset, value); break;
      case 'uint8':   view.setUint8(offset, value); break;
      case 'int16':   view.setInt16(offset, value, true); break;
      case 'uint16':  view.setUint16(offset, value, true); break;
      case 'int32':   view.setInt32(offset, value, true); break;
      case 'uint32':  view.setUint32(offset, value, true); break;
      case 'int64':   view.setBigInt64(offset, BigInt(value), true); break;
      case 'uint64':  view.setBigUint64(offset, BigInt(value), true); break;
      case 'float32': view.setFloat32(offset, value, true); break;
      case 'float64': view.setFloat64(offset, value, true); break;
      case 'pointer': case 'cstring': view.setBigUint64(offset, BigInt(value), true); break;
    }
  }
}

export function struct(fields: Record<string, any>, options?: { packed?: number }): StructClass {
  return createStruct(fields, options);
}

File: packages/senri_ffi_py/src/senri_ffi/struct.py
from __future__ import annotations

import struct as _struct
from typing import Any, Callable

from senri_ffi.pointer import Pointer
from senri_ffi.types.normalized import (
    compute_normalized_layout,
    get_normalized_align,
    get_normalized_size,
    normalize_fields,
)
from senri_ffi.types.constants import TYPE_SIZES


class StructInstance:
    _buffer: bytearray
    _fields: list[dict[str, Any]]
    _struct_cls: Any

    def __init__(self, init: dict[str, Any] | None = None) -> None:
        self._buffer = bytearray(self._struct_cls._total_size)
        if init:
            for key, value in init.items():
                if hasattr(self, key):
                    setattr(self, key, value)

    @property
    def ptr(self) -> Pointer:
        return Pointer({"__buf": self._buffer, "__ptr": 0, "__size": self._struct_cls._total_size})

    def to_pointer(self) -> Pointer:
        alloc_native = getattr(self._struct_cls, "_alloc_native", None)
        if alloc_native:
            native = alloc_native(self._struct_cls._total_size)
            data = native._data if isinstance(native, Pointer) else native
            if data and data.get("__buf"):
                dst = data["__buf"]
                src_bytes = bytes(self._buffer)
                if isinstance(dst, bytearray):
                    dst[:] = src_bytes
                elif hasattr(dst, "_type_") and hasattr(dst, "_length_"):
                    for i, b in enumerate(src_bytes):
                        dst[i] = b
                return native
        return self.ptr


_FMT_MAP: dict[str, str] = {
    "int8": "<b",
    "uint8": "<B",
    "int16": "<h",
    "uint16": "<H",
    "int32": "<i",
    "uint32": "<I",
    "int64": "<q",
    "uint64": "<Q",
    "float32": "<f",
    "float64": "<d",
    "pointer": "<Q",
    "cstring": "<Q",
}


def _read_value(buf: Any, offset: int, type_name: str, _size: int) -> Any:
    fmt = _FMT_MAP.get(type_name)
    if fmt:
        return _struct.unpack_from(fmt, buf, offset)[0]
    return 0


def _write_value(buf: Any, offset: int, type_name: str, _size: int, value: Any) -> None:
    fmt = _FMT_MAP.get(type_name)
    if fmt:
        if type_name in ("pointer", "cstring"):
            value = int(value)
        _struct.pack_into(fmt, buf, offset, value)


def _make_struct_class(
    fields: dict[str, Any],
    packed: int | None = None,
    alloc_native: Callable[[int], Any] | None = None,
    free_native: Callable[[Any], None] | None = None,
) -> type:
    normalized = normalize_fields(fields)
    field_infos: list[dict[str, Any]] = []
    offset = 0
    for name, nt in normalized.items():
        a = min(get_normalized_align(nt), packed) if packed else get_normalized_align(nt)
        if not packed and offset % a != 0:
            offset = ((offset // a) + 1) * a
        type_name = nt.name if nt.kind == "primitive" else None
        field_infos.append({
            "name": name,
            "offset": offset,
            "type": nt,
            "type_name": type_name,
            "size": TYPE_SIZES.get(nt.name, 0) if nt.kind == "primitive" and nt.name is not None else get_normalized_size(nt),
        })
        offset += TYPE_SIZES.get(nt.name, 0) if nt.kind == "primitive" and nt.name is not None else get_normalized_size(nt)

    total_size = offset
    if not packed:
        max_align = max((get_normalized_align(nt) for nt in normalized.values()), default=1)
        total_size = ((offset + max_align - 1) // max_align) * max_align

    cls_dict: dict[str, Any] = {
        "_buffer": bytearray(total_size),
        "_fields": field_infos,
        "_struct_cls": None,
        "_total_size": total_size,
        "_packed": packed,
        "_alloc_native": alloc_native,
        "_free_native": free_native,
    }

    for fi in field_infos:
        fname = fi["name"]
        foff = fi["offset"]
        ftype_name = fi["type_name"]
        fsize = fi["size"]

        def _make_getter(off, tn, sz):
            def getter(self):
                return _read_value(self._buffer, off, tn, sz)
            return getter

        def _make_setter(off, tn, sz):
            def setter(self, value):
                _write_value(self._buffer, off, tn, sz, value)
            return setter

        cls_dict[fname] = property(_make_getter(foff, ftype_name, fsize), _make_setter(foff, ftype_name, fsize))

    def from_pointer(cls, ptr: Any) -> StructInstance:
        inst = object.__new__(cls)
        inst._buffer = bytearray(total_size)
        inst._fields = field_infos
        src_ptr = ptr if isinstance(ptr, Pointer) else Pointer(ptr)
        if src_ptr._data and src_ptr._data.get("__buf"):
            src = src_ptr._data["__buf"]
            if isinstance(src, bytearray):
                inst._buffer[:] = src[:total_size]
            else:
                inst._buffer[:] = bytes(src[:total_size])
        return inst

    cls_dict["__senri_type"] = "struct"
    cls_dict["fields"] = fields
    cls_dict["packed"] = packed
    cls_dict["from_pointer"] = classmethod(from_pointer)

    cls = type("StructInstance", (StructInstance,), cls_dict)
    cls._struct_cls = cls
    return cls


_alloc_native: Callable[[int], Any] | None = None
_free_native: Callable[[Any], None] | None = None


def set_struct_alloc(alloc_fn: Callable[[int], Any], free_fn: Callable[[Any], None]) -> None:
    global _alloc_native, _free_native
    _alloc_native = alloc_fn
    _free_native = free_fn


def struct(fields: dict[str, Any], options: dict[str, Any] | None = None) -> type:
    packed = options.get("packed") if options else None
    return _make_struct_class(fields, packed, _alloc_native, _free_native)
